test/src/testkernel.ts

```diff
@@ -1245,6 +1246,491 @@ describe('jupyter.services - kernel', () => {
+              tester.send(msgDone);
+            };
+
+            let calls: string[] = [];
+            kernel.registerMessageHook(parent_header.msg_id, (msg) => {
+              calls.push('last');
+              kernel.registerMessageHook(parent_header.msg_id, (msg) => {
+                calls.push('first');
+                return true;
+              });
+              return true;
+            });
+
+            future.onIOPub = () => {
+              calls.push('iopub')
+            };
+
+            future.onDone = () => {
+              expect(calls).to.eql(['last', 'iopub', 'first', 'last', 'iopub']);
+              doLater(() => {
+                done();
+              });
+            };
+          });
+        });
+      });
+
+      it('should deactivate a hook immediately on removal', (done) => {
+        let tester = new KernelTester();
+        createKernel(tester).then(kernel => {
+          let options: KernelMessage.IExecuteRequest = {
+            code: 'test',
+            silent: false,
+            store_history: true,
+            user_expressions: {},
+            allow_stdin: false,
+            stop_on_error: false
+          };
+          let future = kernel.execute(options, false);
+          tester.onMessage((message) => {
+            // send a reply
+            let parent_header = message.header;
+            let msg = createMsg('shell', parent_header);
+            tester.send(msg);
+
+            future.onReply = () => {
+              // trigger onIOPub with a 'stream' message
+              let msgStream = createMsg('iopub', parent_header);
+              msgStream.header.msg_type = 'stream';
+              msgStream.content = { 'name': 'stdout', 'text': 'foo' };
+              tester.send(msgStream);
+              // trigger onDone
+              let msgDone = createMsg('iopub', parent_header);
+              msgDone.header.msg_type = 'status';
+              (msgDone as KernelMessage.IStatusMsg).content.execution_state = 'idle';
+              tester.send(msgDone);
+            };
+
+            let calls: string[] = [];
+            let toDelete = (msg: KernelMessage.IIOPubMessage) => {
+              calls.push('delete');
+              return true;
+            }
+            let toDeleteHook = kernel.registerMessageHook(parent_header.msg_id, toDelete);
+
+            kernel.registerMessageHook(parent_header.msg_id, (msg) => {
+              if (calls.length > 0) {
+                // delete the hook the second time around
+                toDeleteHook.dispose();
+              }
+              calls.push('first');
+              return true;
+            });
+
+            future.onIOPub = () => {
+              calls.push('iopub')
+            };
+
+            future.onDone = () => {
+              expect(calls).to.eql(['first', 'delete', 'iopub', 'first', 'iopub']);
+              doLater(() => {
+                done();
+              });
+            };
```

Comment:
Just as a note, we typically add a line between the closing of the test functions as well.